core-primitives/stf-state-handler/src/state_handler.rs

```diff
@@ -0,0 +1,172 @@
+/*
+	Copyright 2021 Integritee AG and Supercomputing Systems AG
+
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+		http://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+
+*/
+
+#[cfg(feature = "sgx")]
+use std::sync::{SgxRwLock as RwLock, SgxRwLockWriteGuard as RwLockWriteGuard};
+
+#[cfg(feature = "std")]
+use std::sync::{RwLock, RwLockWriteGuard};
+
+use crate::{
+	error::{Error, Result},
+	handle_state::HandleState,
+	query_shard_state::QueryShardState,
+	state_snapshot_repository::VersionedStateAccess,
+};
+use itp_types::ShardIdentifier;
+use std::vec::Vec;
+
+/// Implementation of the `HandleState` trait using global files and locks.
+///
+/// For each call it will make a file access and encrypt/decrypt the state from file I/O.
+/// The lock it uses is therefore an 'empty' dummy lock, that guards against concurrent file access.
+pub struct StateHandler<Repository> {
+	state_snapshot_repository: RwLock<Repository>,
+}
```

Comment:
The `StateHandler` is now not using a global, empty `RwLock` anymore. Instead, it has a lock around the snapshot repository to guard against concurrent access.